### ISSUE TITLE
[Bugs] Re-association starvation with 2 connected UPFs

### ISSUE THREAD

**Issue by LaumiH** [free5gc/free5gc#502]
Hi,

I am working with free5GC using two UPFs currently. The problem I am describing now is kind of vague at the moment, and I hope to investigate it further with your help.

When a UPF fails, the association thread detects this failure through the heartbeat. The UPF then is re-associated.

## Describe the bug
The problem I encounter very frequently is that this re-association actually never happens. I can see the SMF sends the association requests to the UPF, and I can also see the UPF correctly processes them. Yet, the reply is never processed by the SMF. Instead, the SMF fails with the error "Request Transaction [14439]: retry-out after 3 attempt(s)".

This happens even without any UE or RAN connected (I use ueransim). I think it is important to mention that I use a heartbeat interval of 10ms for my tests (with 0 retries and a 10ms timeout, I parametrized [these values](https://github.com/free5gc/pfcp/blob/main/transaction.go#L46)), as I depend on detecting failures fast. 

I do not know yet what causes this issue, but I think it may be connected to the custom PFCP library using by the SMF and UPF. 

## To Reproduce
1. Use latest free5GC v3.3.0.
1. Use standard configuration, insert both UPFs into SMF config.
1. Configure the heartbeat interval to 10ms and change the PFCP retry to 0 and the timeout to 10ms [here](https://github.com/free5gc/pfcp/blob/main/transaction.go#L46). Recompile the SMF with these new values. Or parametrize them and read them from a config file.
1. Start free5GC with two UPFs (I use OpenvSwitch and Linux network namespaces to run both on the same VM).

This is the hard part now, as the problem does not show up always. This is why I suspect it is a race condition or high load related problem. Essentially, wait until one heartbeat failure is falsely detected (on my system, a 10ms heartbeat leads to many false-positive heartbeat misses). You will notice the problem showing up through repeated association failures on the SMF and "new node" messages on the UPF.

## Expected behavior
The association should be successful.

## Screenshots
This is the log output:
2023-10-29T19:14:23.254916858Z [INFO][SMF][Main] Send an Association Setup Request to UPF[10.0.0.12]
2023-10-29T19:14:35.257360226Z [WARN][SMF][PFCP] Failed to setup an association with UPF[10.0.0.12], error:Request Transaction [12038]: retry-out after 3 attempt(s)
2023-10-29T19:14:35.257509751Z [INFO][SMF][Main] Send an Association Setup Request to UPF[10.0.0.12]
2023-10-29T19:14:47.260214657Z [WARN][SMF][PFCP] Failed to setup an association with UPF[10.0.0.12], error:Request Transaction [13238]: retry-out after 3 attempt(s)
2023-10-29T19:14:47.260343311Z [INFO][SMF][Main] Send an Association Setup Request to UPF[10.0.0.12]
2023-10-29T19:14:59.264166054Z [WARN][SMF][PFCP] Failed to setup an association with UPF[10.0.0.12], error:Request Transaction [14439]: retry-out after 3 attempt(s)
2023-10-29T19:14:59.264260957Z [INFO][SMF][Main] Send an Association Setup Request to UPF[10.0.0.12]
2023-10-29T19:15:11.267275581Z [WARN][SMF][PFCP] Failed to setup an association with UPF[10.0.0.12], error:Request Transaction [15637]: retry-out after 3 attempt(s)
2023-10-29T19:15:11.267372434Z [INFO][SMF][Main] Send an Association Setup Request to UPF[10.0.0.12]
2023-10-29T19:15:23.269838515Z [WARN][SMF][PFCP] Failed to setup an association with UPF[10.0.0.12], error:Request Transaction [16837]: retry-out after 3 attempt(s)

And the UPF:
2023-10-29T19:15:29.270721263Z [INFO][UPF][PFCP][LAddr:10.0.0.12:8805] handleAssociationSetupRequest
2023-10-29T19:15:29.270816948Z [INFO][UPF][PFCP][LAddr:10.0.0.12:8805] delete node: Remote node {
  Id: 10.0.0.1
  Address: 10.0.0.1:8805
  Local node: &{sess:[<nil> <nil> <nil> <nil> <nil>] free:[1 2 3 4 5]}
}
2023-10-29T19:15:29.270860139Z [INFO][UPF][PFCP][LAddr:10.0.0.12:8805][CPNodeID:10.0.0.1] New node

I noticed that the heartbeat implementation is not very mature at this moment. So it might be that this is simply a feature that was not tested much before. Please let me know how I can help to further investigate!

I omitted the rest of the bug report template, as the bug is kind of vague and I feel no additional info would help at this moment.

**Comment by LaumiH** [comment]
In this pacp file you can see the association requests and responses from UPF 10.0.0.11, which previously failed. You can see that the SMF continues sending heartbeats to UPF 10.0.0.12, and tries to re-associate the failed UPF. However, although the association requests are correctly processed by the UPF, the responses are not processed by the SMF, and there is no re-association.
[starved_association.pcap.zip](https://github.com/free5gc/free5gc/files/13203021/starved_association.pcap.zip)

Hint: a re-association attempt starts at packet 2113, for example.

I let this process run for 10 minutes and there was no re-association.

**Comment by LaumiH** [comment]
Here is a screenshot of the logs (I have adapted the log messages, so this is not the standard output):
![starved_association_5](https://github.com/free5gc/free5gc/assets/37472890/a8e73468-1ffb-4fb2-baeb-ec07648e6ebf)

This is also a better pcap file showing just the relevant association messages on the N4 interface. The heartbeat at that time is stopped completely on both UPFs, so the SMF really only sends the association setup messages. No UEs are connected.
You can see that the retry mechanism indeed seems to work, a retry is sent every 3 seconds. However, the SMF does not receive the responses from the transactions' `EventChannel`.
[starved_association_5.pcap.zip](https://github.com/free5gc/free5gc/files/13205236/starved_association_5.pcap.zip)
![starved_association](https://github.com/free5gc/free5gc/assets/37472890/785a26b6-dc27-4721-a31e-d9d1803de754)

To me, it seems like the transaction channel is closed or read by some other transaction instead.

**Comment by LaumiH** [comment]
I think I just found the bug: It is [this operation](https://github.com/free5gc/pfcp/blob/main/pfcpUdp/udp.go#L115) that blocks indefinitely. There is no error thrown however, so the PFCP server receive goroutine just blocks indefinitely.

I would suggest to fix this problem using a timer like this:
```Golang
select {
case tx.EventChannel <- pfcp.ReceiveEvent{
	Type:       pfcp.ReceiveEventTypeValidResponse,
	RemoteAddr: msg.RemoteAddr,
	RcvMsg:     pfcpMsg,
}:
case <-time.After(2 * time.Second):
	logger.PfcpLog.Errorf("Worker %d timed out on sending to EventChannel", id)
}
```

This allows the goroutine to continue if the channel is closed. Maybe the problem is that the channel is closed in the middle of writing to it, I cannot say for sure. But this fixes my problem.

I am going to create a PR with this fix.

**Comment by LaumiH** [comment]
See https://github.com/free5gc/pfcp/pull/16 for a fix to my issue :)

**Comment by tzuchiehhh** [comment]
Hi @LaumiH, 

I am interested in this issue, but I have encountered some problems while attempting to reproduce the scenario. Are both UPFs designated as PSA-UPF, or is there one PSA-UPF and one I-UPF? Does each UPF connect to a separate SMF, or do both UPFs connect to a single SMF? 

Thanks for your assistance

**Comment by LaumiH** [comment]
Hi @tzuchiehhh,

both the UPFs are PSAs, I have no i-UPF in my setup.

Essentially, I found that the heartbeat implementation altogether lacks accuracy and is influenced a lot by the session management process. I have a fix for this, introducing worker threads to the PFCP server of the SMF to handle incoming PFCP messages. Also, I noticed that the SMF and UPF do not use the same PFCP implementation, which is sth. I don't understand yet. Maybe this has historic reasons.

To prove my point, I have created graphs that show the heartbeat behavior of the original implementation and my improvements. All figures have 1 UPF connected, 150 UERANSIM UEs establish a session each, and the heartbeat interval is at 10ms with a 10ms timeout.

First of all, the figure below shows the heartbeat RTT, so the time between a request from the SMF and a received response from the UPF. Notice the enormous variance of the RTT, which reaches even more than 25ms once. Also, the session establishments of the 150 UEs are visible from the period with increased delay at the beginning.
![10ms_150_ues_outliers_rtt](https://github.com/free5gc/free5gc/assets/37472890/1df5aeec-a3f6-446a-bb04-73554bc8b759)

I also took a look at the distance between consecutive requests, and this explains the variance in RTT a lot. I mean, this is a 10ms heartbeat interval, and the distance between consecutive requests should fluctuate around these 10ms. Yet, the figure below contradicts this. It indicates that heartbeats are sent whenever the SMF wants to (or more likely is able to). Cumulative delays lead to very low accuracy of the heartbeat. I can explain the reason with the implementation of the heartbeat: The interval timer is only started when the response to a heartbeat request is received by the SMF, which means that it is super random and nowhere near the expected 10ms. The heartbeats are not started in a thread, but sequentially after receiving the response.
![10ms_150_ues_outliers_req](https://github.com/free5gc/free5gc/assets/37472890/6949f91c-441d-4a52-b8b2-d7205f217320)

Now, the following figures show the behavior with my improvements. Still, the RTT is not as accurate as I hoped it to be, but I think it might also be the UPF's fault, because its PFCP server has the same flaws as the SMF and cannot process parallel requests, I guess.
![10ms_150_ues_outliers_rtt](https://github.com/free5gc/free5gc/assets/37472890/04432d0a-6dd9-434f-b37e-9088af9da923)

The biggest difference is the distance between consecutive requests, which now finally fluctuates around 10ms with my improvements.
![10ms_150_ues_outliers_req](https://github.com/free5gc/free5gc/assets/37472890/8843d026-6570-4d08-b8a5-2d29bfa7fcb6)

I am working on this currently, and if I have time I can open a PR to fix the heartbeat accuracy.

**Comment by LaumiH** [comment]
Hi @tzuchiehhh,

both the UPFs are PSAs, I have no i-UPF in my setup.

Essentially, I found that the heartbeat implementation altogether lacks accuracy and is influenced a lot by the session management process. I have a fix for this, introducing worker threads to the PFCP server of the SMF to handle incoming PFCP messages. Also, I noticed that the SMF and UPF do not use the same PFCP implementation, which is sth. I don't understand yet. Maybe this has historic reasons.

**Comment by tim-ywliu** [comment]
@LaumiH 

> Also, I noticed that the SMF and UPF do not use the same PFCP implementation, which is sth. I don't understand yet. Maybe this has historic reasons.

Yes, upf had been refactored. We're going to refactor smf as upf and replace pfcp with go-pfcp.

**Comment by tzuchiehhh** [comment]
Hi, @LaumiH,
I configured the heartbeat interval to 10ms in smf config file and set NumOfResend=0, ResendRequestTimeOutPeriod =10 and ResendResponseTimeOutPeriod = 10 in pfcp transaction.go. However, repeated association failures didn't happen.  I've noticed that there is no "new node" message on the UPF. Is there anything I might be missing?

![transaction](https://github.com/free5gc/free5gc/assets/135809479/da4a12fa-e1c5-4b4d-acf8-03b5f78a7375)

![smf](https://github.com/free5gc/free5gc/assets/135809479/f474774f-99b6-491e-81f4-0e2e291cd733)

![upf](https://github.com/free5gc/free5gc/assets/135809479/c4c353b3-6573-4336-bdf4-adf0d0bc5384)

Thanks for your assistance

**Comment by LaumiH** [comment]
Hi,

I investigated more into the problem. In fact, I did make changes to the heartbeat behavior previously (and forgot about them ...), because I noticed its inaccuracy. I am sending heartbeats in a separate thread, decoupled from receiving the actual reply from the UPF. 
The way it is implemented currently, a new heartbeat is only sent after receiving a reply from the UPF, which causes the heartbeat interval to fluctuate above the configured interval, e.g. 12ms instead of 10ms.

You have no new node message in the UPF this was something I added for debugging purposes.

If you are currently refactoring the PFCP behavior of the SMF anyways, I might wait before issuing a PR with my changes/ proposal, so as not to have double work in the end :)

Best,
LaumiH
